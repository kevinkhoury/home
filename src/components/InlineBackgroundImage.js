import React from 'react'
import Img from 'gatsby-image'
import PropTypes from 'prop-types'
import { isValidRGBA } from '../utils'

const InlineBackgroundImage = ({
  extraClasses,
  padding,
  justifyContent,
  fluid,
  src,
  alt,
  styles,
  overlayColor,
  children,
}) => {
  const basename = 'inline-bg-img'
  const customStyles = styles || {}

  const defaultStyles = {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    ...customStyles,
  }

  const containerStyles = {
    padding: padding ? padding : 'inherit',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: justifyContent ? justifyContent : 'center',
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  }

  const image =
    typeof src === 'object' && src.childImageSharp ? (
      <Img
        className={`${basename}-image`}
        fluid={src.childImageSharp.fluid}
        alt={alt}
        style={defaultStyles}
      />
    ) : !!fluid ? (
      <Img
        className={`${basename}-image`}
        fluid={fluid}
        alt={alt}
        style={defaultStyles}
      />
    ) : (
      <img
        className={`${basename}-image`}
        alt={alt}
        src={src}
        style={defaultStyles}
      />
    )

  return (
    <div className={basename} style={{ overflow: 'hidden' }}>
      <div className={`${basename}-underlay-wrapper`}>
        {image}
        <div
          className="is-overlay"
          style={{
            backgroundColor: isValidRGBA(overlayColor)
              ? overlayColor
              : 'rgba(0,0,0,0.65)',
          }}
        />
      </div>
      <div
        className={`is-overlay ${extraClasses ? extraClasses : ''}`}
        style={containerStyles}
      >
        {children}
      </div>
    </div>
  )
}

InlineBackgroundImage.propTypes = {
  extraClasses: PropTypes.string,
  padding: PropTypes.string,
  justifyContent: PropTypes.string,
  src: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      childImageSharp: PropTypes.object,
    }),
  ]),
  fluid: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  alt: PropTypes.string.isRequired,
  styles: PropTypes.object,
  overlayColor: PropTypes.string,
  children: PropTypes.node.isRequired,
}

export default InlineBackgroundImage
