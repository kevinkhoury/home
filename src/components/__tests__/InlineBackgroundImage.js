import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import InlineBackgroundImage from '../InlineBackgroundImage'

const DEFAULT_PROPS = {
  src: {
    childImageSharp: {
      fluid: {},
    },
  },
  alt: 'alttext',
  children: <div>children</div>,
}

jest.mock('gatsby-image', () => (props) => (
  <div data-testid="gatsby-image">{JSON.stringify(props)}</div>
))

afterEach(() => {
  jest.clearAllMocks()
})

describe('InlineBackgroundImage', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<InlineBackgroundImage {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('accepts a childImageSharp for src and converts it to a fluid prop', () => {
    const props = { ...DEFAULT_PROPS }
    render(<InlineBackgroundImage {...props} />)
    const elem = screen.getByTestId('gatsby-image')
    expect(elem.innerHTML).toMatch(/"fluid":{}/)
  })
  it('accepts a string for src', () => {
    const props = { ...DEFAULT_PROPS, src: '/foobar.jpg' }
    render(<InlineBackgroundImage {...props} />)
    const elem = screen.getByAltText('alttext')
    expect(elem.getAttribute('src')).toEqual('/foobar.jpg')
  })
  it('accepts a fluid prop instead of src', () => {
    const props = { alt: 'alttext', children: <div>children</div>, fluid: {} }
    render(<InlineBackgroundImage {...props} />)
    const elem = screen.getByTestId('gatsby-image')
    expect(elem.innerHTML).toMatch(/"fluid":{}/)
  })
  it('prefers a childImageSharp src over a fluid prop', () => {
    const props = { ...DEFAULT_PROPS, fluid: { src: 'not me' } }
    render(<InlineBackgroundImage {...props} />)
    const elem = screen.getByTestId('gatsby-image')
    expect(elem.innerHTML).not.toMatch(/"not me"/)
  })
  it('accepts an array for the fluid prop', () => {
    const props = {
      alt: 'alttext',
      children: <div>children</div>,
      fluid: ['fluid1', 'fluid2'],
    }
    render(<InlineBackgroundImage {...props} />)
    const elem = screen.getByTestId('gatsby-image')
    expect(elem.innerHTML).toMatch(/"fluid1","fluid2"/)
  })
  it('accepts a custom rgba value for the overlayColor if well-formed', () => {
    const propsValid = {
      ...DEFAULT_PROPS,
      overlayColor: 'rgba(255, 255,      255,0.35)',
    }
    const propsInvalid = {
      ...DEFAULT_PROPS,
      overlayColor: 'foobar',
    }
    render(
      <div>
        <div data-testid="valid">
          <InlineBackgroundImage {...propsValid} />
        </div>
        <div data-testid="invalid">
          <InlineBackgroundImage {...propsInvalid} />
        </div>
      </div>,
    )
    const valid = screen.getByTestId('valid').querySelector('.is-overlay')
    const invalid = screen.getByTestId('invalid').querySelector('.is-overlay')
    expect(valid.style.backgroundColor).toEqual('rgba(255, 255, 255, 0.35)')
    expect(invalid.style.backgroundColor).not.toBe('foobar')
  })
  // it('', () =>{  })
  // it('', () =>{  })
  // it('', () =>{  })
  // it('', () =>{  })
  // it('', () =>{  })
  // it('', () =>{  })
  // it('', () =>{  })
  // it('', () =>{  })
})
