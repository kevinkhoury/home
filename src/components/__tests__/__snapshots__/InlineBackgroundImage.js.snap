// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineBackgroundImage snapshot renders properly 1`] = `
<div
  className="inline-bg-img"
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <div
    className="inline-bg-img-underlay-wrapper"
  >
    <div
      data-testid="gatsby-image"
    >
      {"className":"inline-bg-img-image","fluid":{},"alt":"alttext","style":{"position":"absolute","top":0,"bottom":0,"left":0,"right":0}}
    </div>
    <div
      className="is-overlay"
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.65)",
        }
      }
    />
  </div>
  <div
    className="is-overlay "
    style={
      Object {
        "bottom": 0,
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "left": 0,
        "padding": "inherit",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <div>
      children
    </div>
  </div>
</div>
`;
